Fix #20481: Properly deselect items on close

mrmbernardi opened the pull request:
Fix for #20481 where the tile inspector wouldn't visually deselect items when closing.

I also noticed `ToolCancel` is only called when closing via the mouse. I've moved it into the window `OnClose` handler to make it work consistently.

ZehMatt commented on src/openrct2-ui/windows/TileInspector.cpp:
```diff
@@ -712,6 +711,12 @@ class TileInspector final : public Window
+        auto elem = GetSelectedElement();
+        if (elem != nullptr)
+        {
+            MapInvalidateElement(_toolMap, elem);
```
This is a bit problematic in multiplayer, someone else might have deleted the tile element. Its probably better to store the position of the tile where the element was selected and use that to invalidate the tile, the pointer is only used during the rendering path to see if it requires to be highlighted without accessing the actual memory of it.

mrmbernardi replied:
That's already what this code does. `GetSelectedElement()` uses the position of the selected element combined with the selected index to return a pointer.

mrmbernardi replied:
On second thought, the selected index could be greater than the number of elements in the tile if someone deletes stuff on multiplayer. I'm not fully sure, but I think this would break everything that uses the tile inspector selection if it doesn't listen to updates. 

The only solution would be to re-iterate through the tile and count the number of elements every time the selection is used.

ZehMatt replied:
We don't need to know which specific element was selected for invalidation to work, the invalidation is actually quite primitive as it just marks on a fixed size grid which area of the screen requires to be redrawn. Storing a pointer to the tile element to be read at a later point of time is not safe to do.

ZehMatt reviewed (changes requested):
Tested this and it fixes the issue of not invalidating

ZehMatt reviewed (approved):
All looks good, sorry about the confusion.

mrmbernardi commented:
On second thought, I'm not happy about what I mention here: https://github.com/OpenRCT2/OpenRCT2/pull/20482#discussion_r1257512901 

I'd like to expand the scope of the PR and fix the issue properly before merging

ZehMatt commented:
> On second thought, I'm not happy about what I mention here: [#20482 (comment)](https://github.com/OpenRCT2/OpenRCT2/pull/20482#discussion_r1257512901)
> 
> I'd like to expand the scope of the PR and fix the issue properly before merging

There is a function called MapGetNthElementAt which will return nullptr with an invalid index.

mrmbernardi commented:
I've changed `GetSelectedElement()` to use `MapGetNthElementAt()` and also it'll return `nullptr` if there is nothing selected. As a result, I've done a minor refactor to take advantage of this new behaviour and remove an instance where it was looking up the tile element twice. I've also changed the variables holding the return value to be `const TileElement* const`, as I figure we may as well use the `const` keyword since someone bothered to make the functions in `TileElement` support it.

ZehMatt commented:
Is this still relevant given #20591 also fixes this?
